### PR TITLE
Cyclic Spells During Research

### PR DESCRIPTION
**Pull request by drboldwin**
Also a small update to Training_manager to handle releasing cyclic spells during transitions from crossing_training to hunting_buddy

### CODE REVIEW

**Comment by drboldwin** [comment]
Redid this so it was by itself.

File: training-manager.lic [hunk 1]
```diff
@@ -99,6 +102,17 @@ class TrainingManager
       end
     end
   end
+  
+  def check_cyclics
+    echo DRSpells.active_spells
```
Comment (rpherbig) [inline training-manager.lic]: This should be conditional on the debug flag

**Comment by drboldwin** [comment]
Oops - will add

**Comment by drboldwin** [comment]
Tried to do that but would no longer let me likely because the fork I used to create this pull request has been deleted.

Although the line can be deleted if you would like since I don't believe there is a debug flag for Training-manager.  I would likely end up trying to add the crossing-trianing debug but not sure if that would work.  Like this:

`echo DRSpells.active_spells if UserVars.crossing_trainer_debug`

**Comment by drboldwin** [comment]
but really I forgot I had that echo in there and I am addicted to echoing things when I code.  Got rid of most of them and missed this one.

File: training-manager.lic [hunk 2]
```diff
@@ -99,6 +102,17 @@ class TrainingManager
...
+    fput('release gj') if DRSpells.active_spells['Glythtide\'s Joy']
+    fput('release care') if DRSpells.active_spells['Caress of the Sun']
+    fput('release bes') if DRSpells.active_spells['Bear Strength']
+    fput('release fg') if DRSpells.active_spells['Faenella\'s Grace']
+    fput('release rezz') if DRSpells.active_spells['Resurrection']
+    fput('release ghs') if DRSpells.active_spells['Ghost Shroud']
+    return
```
Comment (rpherbig) [inline training-manager.lic]: This return is unnecessary

File: training-manager.lic [hunk 3]
```diff
@@ -99,6 +102,17 @@ class TrainingManager
       end
     end
   end
+  
+  def check_cyclics
```
Comment (rpherbig) [inline training-manager.lic]: This method might belong in common-arcana... what do you think?

File: crossing-training.lic [hunk 4]
```diff
@@ -413,13 +413,22 @@ class CrossingTraining
   end
 
   def research
+    if @settings.cyclic_research
+      research_spell = @settings.crossing_cyclic_research_spells([])
```
Comment (rpherbig) [inline crossing-training.lic]: This appears to be a list of skills, and I feel 'crossing' is not communicating any information. I recommend naming it `cyclic_research_skills`.

File: crossing-training.lic [hunk 5]
```diff
@@ -413,13 +413,22 @@ class CrossingTraining
   end
 
   def research
+    if @settings.cyclic_research
```
Comment (rpherbig) [inline crossing-training.lic]: Is this boolean necessary, or is the presence of `crossing_cyclic_research_spells` (or `cyclic_research_skills`) enough to indicate this should be done?

File: crossing-training.lic [hunk 6]
```diff
@@ -413,13 +413,22 @@ class CrossingTraining
...
+      research_spell = @settings.crossing_cyclic_research_spells([])
+                       .min_by { |s| DRSkill.getxp(s) }
+      cyclic_research(research_spell)
```
Comment (rpherbig) [inline crossing-training.lic]: This appears to work for non-cyclic spells too. Is that intentional? If so, should any of the naming change?

**Comment by rpherbig** [comment]
This is the sample YAML excerpt posted in the previous PR:
```
crossing_cyclic_research_spells:
- Utility
- Warding


cyclic_research: true
research_training_spells:
  Warding:
    abbrev: GHS
    cyclic: true
  Utility:
    abbrev: REZZ
    cyclic: true
```

**Comment by drboldwin** [comment]
Some of these will be addressed when we finish up the Universal Cyclic Release as seen here:

https://github.com/rpherbig/dr-scripts/issues/674

It will take care of the defining of cyclics in training-manager.lic as well as the hardcoded aspect in crossing-training.lic and should be pushed before this is so that we can adjust accordingly.

> This appears to work for non-cyclic spells too. Is that intentional? If so, should any of the naming change?

It does though it was not my intent.  I guess I just didn't account for non-cyclic since the only spells listed under research_training_spells should be cyclic spells anyways.  I was unsure of how to go about first picking the lower ability and then determining if it was cyclic and then going back and picking a different ability if it wasn't.  If this logic is put in then we could avoid listing the spells separately completely and base it off of training_spells as you had previously mentioned.

> This appears to be a list of skills, and I feel 'crossing' is not communicating any information. I recommend naming it cyclic_research_skills.

Agreed here - crossing has no actual bearing to it - I was simply copying the structure used for listing skills to train in crossing which was crossing_training so I called it crossing_cyclic_research_spells.  With the expanding of hometowns and whatnot this seems even more unnecessary.

**Comment by drboldwin** [comment]
Closing this and reopening to reflect new changes to common-arcana

**Pull request closed without merging**